app/services/sign_in/refresh_token_decryptor.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module SignIn
+  class RefreshTokenDecryptor
+    attr_reader :encrypted_refresh_token
+
+    def initialize(encrypted_refresh_token:)
+      @encrypted_refresh_token = encrypted_refresh_token
+    end
+
+    def perform
+      split_refresh_token = split_encrypted_refresh_token(encrypted_refresh_token)
+
+      decrypted_part = get_decrypted_part(split_refresh_token)
+      nonce = get_nonce_part(split_refresh_token)
+      version = get_version_part(split_refresh_token)
+
+      validate_token!(decrypted_part, nonce, version)
+
+      SignIn::RefreshToken.new(session_handle: decrypted_part.session_handle,
+                               user_uuid: decrypted_part.user_uuid,
+                               parent_refresh_token_hash: decrypted_part.parent_refresh_token_hash,
+                               anti_csrf_token: decrypted_part.anti_csrf_token,
+                               nonce: decrypted_part.nonce,
+                               version: decrypted_part.version)
+    end
+
+    private
+
+    def validate_token!(decrypted_token, nonce, version)
+      raise SignIn::Errors::RefreshVersionMismatchError if decrypted_token.version != version
+      raise SignIn::Errors::RefreshNonceMismatchError if decrypted_token.nonce != nonce
+    end
+
+    def get_decrypted_part(split_refresh_token)
+      decrypt_refresh_token(split_refresh_token[SignIn::Constants::REFRESH_TOKEN_ENCRYPTED_POSITION])
+    end
+
+    def get_nonce_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_NONCE_POSITION]
+    end
+
+    def get_version_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_VERSION_POSITION]
+    end
+
+    def split_encrypted_refresh_token(encrypted_refresh_token)
+      encrypted_refresh_token.split('.')
```

Comment:
```suggestion
      encrypted_refresh_token.split('.', 3)
```

Should splitting be limited to three parts?

<img width="404" alt="image" src="https://user-images.githubusercontent.com/28198/156262892-e8743594-64f7-4b07-be17-8f422123910d.png">

Reply:
Only reason I hesitate is to keep the ability to split versatile, with the ability to change the number of fields if necessary with different versions

Reply:
You know, this gives me an idea. Maybe it's appropriate to create another model: `EncryptedRefreshToken` that incapsulates this logic. Would probably make this more straightforward

Reply:
I formatted the constants a bit differently to give me an array length of these elements, so I can use that as a way to limiting the splitting and remain versatile